src/lib/pubkey/sphincsplus/sphincsplus.cpp

```diff
@@ -0,0 +1,421 @@
+/*
+* Sphincs+
+* (C) 2023 Jack Lloyd
+*     2023 Fabian Albert, René Meusel, Amos Treiber - Rohde & Schwarz Cybersecurity
+*
+* Botan is released under the Simplified BSD License (see license.txt)
+**/
+
+#include <botan/sphincsplus.h>
+#include <botan/rng.h>
+
+#include <botan/internal/pk_ops_impl.h>
+#include <botan/internal/sp_fors.h>
+#include <botan/internal/sp_hash.h>
+#include <botan/internal/sp_treehash.h>
+#include <botan/internal/sp_types.h>
+#include <botan/internal/sp_wots.h>
+#include <botan/internal/sp_xmss.h>
+#include <botan/internal/stl_util.h>
+
+#if not defined(BOTAN_HAS_SHAKE) and not defined(BOTAN_CAN_USE_SHA_FOR_SHPINCSPLUS)
+   static_assert(false, "botan module 'sphincsplus' is useful only when enabling at least 'shake' or 'sha2_32,sha2_64,mgf1,hmac,trunc_hash'");
+#endif
```

Comment:
> I'd be inclined here to just hard require everything

That's what we did initially. But I found it a worthwhile use case to build the library with SPHINCS+ but without support for one of the hashes. There isn't much additional complexity to allow for this flexibility, in my opinion.

The static assert was added to provide guidance to users that might configure as `--minimized-build --enable-modules=sphincsplus` and end up with a dysfunctional module. A similar approach was implemented for Kyber (to require at least `kyber` or `kyber_90s` to be enabled). In contrast, the Kyber example uses actual sub modules that can model their respective transitive dependencies. We could do this here as well, maybe.

Reply:
OK. I would somewhat prefer a submodules approach here eg `sphincsplus_sha2` and `sphinxplus_sha3` which themselves depend on the relevant dependencies. That could also provide a path for encapsulating the operation of the hash algorithms, which would probably prove useful if/when Haraka support is added.

Also explicit submodules could expose `BOTAN_HAS_SPHINCSPLUS_SHAn` feature macros which would allow applications to test which hash functions are supported, without having to know about this one off macro.

*And* (if I ever get around to adding back support for conditional module dependencies) we could avoid the static assert by having `sphinxplus` module depend on `sphincsplus_sha2 | sphincsplus_sha3`.

Reply:
Yeah. Plus it is consistent with Kyber and Dilithium where the AES modes were handled as submodules. On it.